### PR TITLE
Add support for gitlab push and system hooks

### PR DESCRIPTION
**Pull request by martindekov**
Adding gitlab-push function which can be added to webhook route
to handle push or system hooks from gitlab

Signed-off-by: Martin Dekov (VMware) <mdekov@vmware.com>

## Description

Handling push-events which deploy your function from repo in git-lab.

## How Has This Been Tested?

1. Deploy the function in your OF-cloud instance `faas-cli deploy -f gitlab.yml`
2. Configure your gitlab webhook route to `<gateway_address>/functions/gitlab-push`
3. Make function somewhere `faas-cli new <funcname> --lang=go --prefix=<your dockerhub name>`
4. ```mv <funcname>.yml stack.yml```
5. Push function to your gitlab repo
6. Check function deployed in `<gateway>/functions/overview?username=<name of account in gitlab>`

## Checklist:

I have:

- [ ] updated the documentation if required
- [x] read the [CONTRIBUTION](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md) guide
- [x] signed-off my commits with `git commit -s`
- [ ] added unit tests

### CODE REVIEW

File: gitlab-push/handler.go [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+
+	"github.com/openfaas/openfaas-cloud/sdk"
+)
+
+const Source = "gitlab-push"
+
+// Handle a serverless request
+func Handle(req []byte) string {
+
+	event := os.Getenv("Http_X_Gitlab_Event")
```
Comment (alexellis) [inline gitlab-push/handler.go]: Can you add a unit test for these scenarios?

Comment (martindekov) [reply gitlab-push/handler.go]: Yeah

File: gitlab-push/handler.go [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+		},
+	}
+
+	var found bool
+
+	if readBool("validate_customers") {
+
+		customersURL := os.Getenv("customers_url")
+
+		customers, getErr := getCustomers(customersURL)
```
Comment (alexellis) [inline gitlab-push/handler.go]: I found a bug in the code in the original version, can you sync up please?

File: gitlab-push/handler.go [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+}
+
+func postEvent(pushEvent sdk.PushEvent) (int, error) {
+	gatewayURL := os.Getenv("gateway_url")
+
+	body, _ := json.Marshal(pushEvent)
+
+	c := http.Client{}
+	bodyReader := bytes.NewBuffer(body)
+	httpReq, _ := http.NewRequest(http.MethodPost, gatewayURL+"async-function/git-tar", bodyReader)
```
Comment (alexellis) [inline gitlab-push/handler.go]: You have to sign this now.. see the GitHub variant.

File: gitlab-push/handler.go (hunk 2 above)
Comment (martindekov) [reply gitlab-push/handler.go]: Should I add it as separate commit to see the difference or amend to this one?

**Comment by alexellis** [comment]
Regarding gitlab-push, do we actually want a gitlab-event so that we can handle other events? What events are available in GitLab?

**Comment by alexellis** [comment]
Part 1 -So - after a planning meeting with the faas team, we want to restrict gitlab-push to only work when repos are tagged as "openfaas-cloud".

Part 2 - We need to garbage collect repos / functions when the tag is removed.
Tag events are sent via System Hooks.

**Comment by martindekov** [comment]
Derek add label: blocked

**Comment by martindekov** [comment]
Also now depends on #286 to be existing in the proper way.

**Comment by martindekov** [comment]
And on #291 to amend fix to send scm field and construct proper stack SCM field since currently it depends on PR #281 .

**Comment by martindekov** [comment]
Also amended changed did in PR #206 into this function.

**Comment by martindekov** [comment]
Also amended change #291 to the function

File: gitlab-push/handler.go [hunk 4]
```diff
@@ -0,0 +1,256 @@
...
+
+	if pushEvent.Ref != "refs/heads/master" {
+		msg := "refusing to build non-master branch: " + pushEvent.Ref
+		auditEvent := sdk.AuditEvent{
+			Message: msg,
+			Owner:   pushEvent.Repository.Owner.Login,
+			Repo:    pushEvent.Repository.Name,
+			Source:  Source,
+		}
+
```
Comment (ivanayov) [inline gitlab-push/handler.go]: Should report status as well

**Comment by martindekov** [comment]
Derek remove label: blocked

**Comment by martindekov** [comment]
Work to be merged in #308

**Pull request closed without merging**